src/math/log_10.c

```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: BSD-3-Clause/
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+// Author: Shriram Shastry <malladi.sastry@linux.intel.com>
+//
+//
+
+#include <sof/math/base2log.h>
+
+/**
+ * Base-10 logarithm  log10(x)
+ *
+ * loge = (u) computes the base-10 logarithm of
+ * u using lookup table.
+ * input u must be scalar/real number and positive
+ *
+ * +------------------+-----------------+--------+--------+
+ * | inpfxp	      |log10(returntype)| inpfxp |  loge  |
+ * +----+-----+-------+----+----+-------+--------+--------+
+ * |WLen| FLen|Signbit|WLen|FLen|Signbit| Qformat| Qformat|
+ * +----+-----+-------+----+----+-------+--------+--------+
+ * | 32 | 0   |	 1    | 32 | 28 | 0     | 32.0	 | 4.28   |
+ * +------------------+-----------------+--------+--------+
+ * Arguments	: uint32_t numerator [1 to 4294967295, Q32.0]
+ * Return Type	: uint32_t UQ4.28 [0 to 9.6329499409]
+ */
+uint32_t log_10_int32(uint32_t numerator)
+{
+	uint32_t logb10fxp;
+
+	logb10fxp = (uint32_t)(((int64_t)base2_logarithm(numerator) << BASE_10_LOG_NUM_FRACTBITS) /
+			       BASE_10_PRECALC_DENOM_10_UQ18P14);
```

Comment:
Could you try to replace this by multiplication?
#define ONE_OVER_LOG2_10_Q1P31 0x.....  /* 1/log2(10) */

logb10fxp = (uint32_t)(((int64_t)base2_logarithm(numerator) * ONE_OVER_LOG2_10_Q1P31) >> (16+31-28));

In theory a multiplication op should be faster than a division op.

Reply:
Replacing the division by multiplication involve few additional hurdles. I could see 

`DIVISION`          CMP_TOLERANCE 0.0000122948805111	Error[max] = 0.0000122948805111,THD = -98.2055137526933
`MULTIPLICATION` CMP_TOLERANCE 0.0001528374565112	Error[max] = 0.0001528374565112,THD = -76.3154039672191828

I can't reuse `base2_logarithm()`- 

As expected cpu cycle per sample for multiplication has reduced by 10% compared to division
One can make lookup table common across base`2`, base`10 `and base`e`
```
int64_t base_10_log(uint32_t numerator)
{
        static const int32_t iv1[129] = {
            0,     736,   1466,  2190,  2909,  3623,  4331,  5034,  5732,  6425,  7112,  7795,
            8473,  9146,  9814,  10477, 11136, 11791, 12440, 13086, 13727, 14363, 14996, 15624,
            16248, 16868, 17484, 18096, 18704, 19308, 19909, 20505, 21098, 21687, 22272, 22854,
            23433, 24007, 24579, 25146, 25711, 26272, 26830, 27384, 27936, 28484, 29029, 29571,
            30109, 30645, 31178, 31707, 32234, 32758, 33279, 33797, 34312, 34825, 35334, 35841,
            36346, 36847, 37346, 37842, 38336, 38827, 39316, 39802, 40286, 40767, 41246, 41722,
            42196, 42667, 43137, 43603, 44068, 44530, 44990, 45448, 45904, 46357, 46809, 47258,
            47705, 48150, 48593, 49034, 49472, 49909, 50344, 50776, 51207, 51636, 52063, 52488,
            52911, 53332, 53751, 54169, 54584, 54998, 55410, 55820, 56229, 56635, 57040, 57443,
            57845, 58245, 58643, 59039, 59434, 59827, 60219, 60609, 60997, 61384, 61769, 62152,
            62534, 62915, 63294, 63671, 64047, 64421, 64794, 65166, 65536};
        static const int8_t iv[256] = {
            8, 7, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
            2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
        
        int32_t b;
        uint32_t v;
        uint32_t xfi;
        int num_left_shifts;
        int shift_factor;
        unsigned int slice_temp;
        /*compute numerator*/
        shift_factor = iv[(numerator >> 24ULL)];
        v = numerator << shift_factor;
        num_left_shifts = shift_factor;

        shift_factor = iv[v >> 24ULL];
        v <<= shift_factor;
        num_left_shifts += shift_factor;

        shift_factor = iv[v >> 24ULL];
        v <<= shift_factor;
        num_left_shifts += shift_factor;

        shift_factor = iv[v >> 24ULL];
        xfi = v << shift_factor;
        num_left_shifts += shift_factor;

        slice_temp = ((v << shift_factor) >> 24ULL); 

        b = 19728; /* denominator const for term 1 / log2(10)*/
        
        int32_t a1 = (31 - num_left_shifts);
        a1 <<= 16;
        int32_t a2 = iv1[((xfi >> 24ULL) - 127U) - 1];
        int32_t x_i = a1 + a2;
        int32_t b_1 = iv1[255 & (slice_temp + 130U) - 1];
        int32_t b_2 = iv1[(255 & (slice_temp + 130U) - 1U) - 1];
        int32_t y_i = b_1 - b_2;
        /*Product = numerator * denominator */
        return (x_i + ((((xfi & 16777215U) * (int64_t)y_i) >> 24ULL))) * b;
```

Reply:
So you are saying that the error is much bigger by multiplication(1) than division(2)?
(1)
```
#define ONE_OVER_LOG2_10_Q1P31 0x268826A2 /* 1/log2(10) */
logb10fxp = (uint32_t)(((int64_t)base2_logarithm(numerator) * ONE_OVER_LOG2_10_Q1P31) >> (16+31-28));
```
(2)
```
#define LOG2_10_UQ16P16 0x3526A /* log2(10) */
logb10fxp = (uint32_t)(((int64_t)base2_logarithm(numerator) << (16-16+28)) / LOG2_10_UQ16P16);
```

It's a surprising result for me however it is possible (Magics of fractional bit accuracy.) If the error is not acceptable for the multiplication method then let's keep the division method.

Reply:
Pls check new implementation ( I am using multiplication and we have higher accuracy  without performance drop)